cms/djangoapps/modulestore_migrator/models.py

```diff
@@ -0,0 +1,217 @@
+"""
+Models for the modulestore migration tool.
+"""
+from __future__ import annotations
+
+from django.contrib.auth import get_user_model
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+from user_tasks.models import UserTaskStatus
+
+from model_utils.models import TimeStampedModel
+from opaque_keys.edx.django.models import (
+    LearningContextKeyField,
+    UsageKeyField,
+)
+from openedx_learning.api.authoring_models import (
+    LearningPackage, PublishableEntity, Collection, DraftChangeLog, DraftChangeLogRecord
+)
+
+from .data import CompositionLevel, RepeatHandlingStrategy
+
+User = get_user_model()
+
+
+class ModulestoreSource(models.Model):
+    """
+    A legacy learning context (course or library) which can be a source of a migration.
+    """
+    key = LearningContextKeyField(
+        max_length=255,
+        unique=True,
+        help_text=_('Key of the content source (a course or a legacy library)'),
+    )
+    forwarded = models.OneToOneField(
+        'modulestore_migrator.ModulestoreMigration',
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        help_text=_('If set, the system will forward references of this source over to the target of this migration'),
+        related_name="forwards",
+    )
+
+    def __str__(self):
+        return f"{self.__class__.__name__}('{self.key}')"
+
+    __repr__ = __str__
+
+
+class ModulestoreMigration(models.Model):
+    """
+    Tracks the action of a user importing a Modulestore-based course or legacy library into a
+    learning-core based learning package
+
+    Notes:
+    * As of Ulmo, a learning package is always associated with a v2 content library, but we
+      will not bake that assumption into this model)
+    * Each Migration is tied to a single UserTaskStatus, which connects it to a user and
+      contains the progress of the import.
+    * A single ModulestoreSource may very well have multiple ModulestoreMigrations; however,
+      at most one of them with be the "authoritative" migration, as indicated by `forwarded`.
+    """
+
+    ## MIGRATION SPECIFICATION
+    source = models.ForeignKey(
+        ModulestoreSource,
+        on_delete=models.CASCADE,
+        related_name="migrations",
+    )
```

Comment:
@cmltaWt0 , I just realized that it would be very useful to have a `source_version` CharField here so that we can know what version a course/library was when it was migrated. For legacy libraries in particular, it would be good to know the source version so that we can know whether a "sync" is available for migrated legacy library references. Could you add that field and populate it during the migration process?

Reply:
Absolutely! Great idea.